contracts/CourtPresaleActivate.sol

```diff
@@ -51,8 +57,72 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
 
+        _buyAndActivate(_from, _amount, _token);
+    }
+
+    /**
+    * @dev This function needs a previous approval on the external token used for the contributed amount.
+    *      It will pull the approved tokens, convert them in Uniswap to the presale contribution token,
+    *      convert the obtained tokens in the presale instance, and activate the converted tokens in the
+    *      jurors registry instance of the Aragon Court.
+    * @param _token Address of the external contribution token used
+    * @param _from Address of the original caller (juror) converting and activating the tokens
+    * @param _amount Amount of contribution tokens to be converted and activated
+    * @param _minTokens Minimum amount of presale contribution tokens obtained in Uniswap
+    * @param _minEth Minimum amount of ETH obtained in Uniswap (Uniswap internally converts first to ETH and then to target token)
+    * @param _deadline Transaction deadline for Uniswap
+    */
+    function contributeExternalToken(
+        address _token,
+        address _from,
+        uint256 _amount,
+        uint256 _minTokens,
+        uint256 _minEth,
+        uint256 _deadline
+    )
+        external
+    {
+        require(_amount > 0, ERROR_ZERO_AMOUNT);
+
+        // move tokens to this contract
+        require(ERC20(_token).safeTransferFrom(_from, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
+
+        // get the Uniswap exchange for the contribution token
+        address contributionTokenAddress = address(presale.contributionToken());
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapFactory.getExchange(contributionTokenAddress));
+
+        // swap tokens
+        uint256 contributionTokenAmount = uniswapExchange.tokenToTokenSwapInput(_amount, _minTokens, _minEth, _deadline, _token);
+
+        // buy in presale
+        _buyAndActivate(_from, contributionTokenAmount, contributionTokenAddress);
+    }
+
+    /**
+    * @dev It will send the received ETH to Uniswap to get presale contribution tokens,
+    *      convert the obtained tokens in the presale instance, and activate the converted tokens in the
+    *      jurors registry instance of the Aragon Court.
+    * @param _minTokens Minimum amount of presale contribution tokens obtained in Uniswap
+    * @param _deadline Transaction deadline for Uniswap
+    */
+    function contributeEth(uint256 _minTokens, uint256 _deadline) external payable {
+        require(msg.value > 0, ERROR_ZERO_AMOUNT);
+
+        address contributionTokenAddress = address(presale.contributionToken());
+
+        // get the Uniswap exchange for the contribution token
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapFactory.getExchange(contributionTokenAddress));
+
+        // swap tokens
+        uint256 contributionTokenAmount = uniswapExchange.ethToTokenSwapInput.value(msg.value)(_minTokens, _deadline);
+
```

Comment:
⚠️ Is there any chance that we will end up with leftover ETH that we need to return to `msg.sender`?

Reply:
This is done. I didn't use the "withdrawal pattern" because I don't think it's a good idea (I think that if there are leftovers it's good to revert if they cannot be sent back), but let me know if you think the opposite.